components/fxa-client/src/state_machine/mod.rs

```diff
@@ -0,0 +1,169 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! FxA state machine
+//!
+//! This presents a high-level API for logging in, logging out, dealing with authentication token issues, etc.
+
+use std::sync::Arc;
+
+use error_support::{breadcrumb, convert_log_report_error, handle_error};
+use parking_lot::Mutex;
+
+use crate::{internal, ApiResult, DeviceConfig, Error, FirefoxAccount, Result};
+mod checker;
+mod logic;
+mod types;
+
+pub use checker::FxaStateMachineChecker;
+pub use types::{Event, FxaEvent, FxaState, State};
+
+use logic::next_state;
+
+/// FxA state machine
+///
+/// This provides a high-level interface for using a [FirefoxAccount] -- login, logout, checking
+/// auth status, etc.
+pub struct FxaStateMachine {
+    account: Arc<FirefoxAccount>,
+    state: Mutex<FxaState>,
+    oauth_entrypoint: String,
+    device_config: DeviceConfig,
+}
+
+impl FxaStateMachine {
+    /// Create an FxaStateMachine
+    ///
+    /// Note: When restoring a connected account, only `device_config.capabilities` will be used.
+    /// We will use the device type and device name info that's stored on the server.
+    pub fn new(
+        account: Arc<FirefoxAccount>,
+        oauth_entrypoint: String,
+        device_config: DeviceConfig,
```

Comment:
Maybe we should input the list of OAuth scopes here as well rather than store them in the events.  I didn't do that in case we want to support clients that want flexible permission handling -- for example, they request one set of scopes initially, then upgrade to a new set of scopes later on.